tests/h/services/url_migration_test.py

```diff
@@ -0,0 +1,227 @@
+        )
+
+    def test_move_annotations_by_url_moves_matching_annotations(
+        self,
+        db_session,
+        factories,
+        pyramid_request,
+        update_annotation,
+        move_annotations_task,
+    ):
+        anns = [
+            factories.Annotation(target_uri="https://example.com"),
+            factories.Annotation(target_uri="https://example.com"),
+            factories.Annotation(target_uri="https://othersite.com"),
+            factories.Annotation(target_uri="https://example.com"),
+        ]
+        db_session.flush()
+
+        svc = URLMigrationService(pyramid_request)
+        svc.move_annotations_by_url(
+            "https://example.com",
+            {"url": "https://example.org"},
+        )
+
+        # First annotation should be moved synchronously.
+        assert update_annotation.call_count == 1
+        update_annotation.assert_called_with(
+            pyramid_request,
+            Any.of([a.id for a in anns]),
+            {"target_uri": "https://example.org"},
+            update_timestamp=False,
+            reindex_tag="URLMigrationService.move_annotations",
+        )
+        pyramid_request.tm.commit.assert_called_once()
+
+        moved_ann_id = update_annotation.call_args[0][1]
+        remaining_ann_ids = [
+            a.id
+            for a in anns
+            if a.target_uri == "https://example.com" and a.id != moved_ann_id
+        ]
+
+        # Remaining matching annotations should be moved in separate tasks.
+        assert move_annotations_task.delay.call_count == 1
+        move_annotations_task.delay.assert_called_with(
+            Any.list.containing(remaining_ann_ids).only(),
+            "https://example.com",
+            {"url": "https://example.org"},
+        )
+
+    def test_move_annotations_by_url_handles_no_matches(
+        self,
+        db_session,
+        factories,
+        pyramid_request,
+        update_annotation,
+        move_annotations_task,
+    ):
+        # Make sure there are some non-matching annotations in the DB.
+        factories.Annotation(target_uri="https://foo.com")
+        factories.Annotation(target_uri="https://bar.com")
+        factories.Annotation(target_uri="https://baz.com")
+        db_session.flush()
+
+        svc = URLMigrationService(pyramid_request)
+        svc.move_annotations_by_url(
+            "https://example.com",
+            {"url": "https://example.org"},
+        )
+
+        update_annotation.assert_not_called()
+        move_annotations_task.delay.assert_not_called()
+
+    @pytest.fixture
+    def transform_document(self, patch):
+        return patch("h.services.url_migration.transform_document")
+
+    @pytest.fixture
+    def update_annotation(self, patch):
+        return patch("h.storage.update_annotation")
+
+    @pytest.fixture
+    def move_annotations_task(self, patch):
+        return patch("h.services.url_migration.move_annotations")
```

Comment:
I'd make these three autouse fixtures just because we don't want any unit tests to accidentally be calling the real code:

```suggestion
    @pytest.fixture(autouse=True)
    def transform_document(self, patch):
        return patch("h.services.url_migration.transform_document")

    @pytest.fixture(autouse=True)
    def update_annotation(self, patch):
        return patch("h.storage.update_annotation")

    @pytest.fixture(autouse=True)
    def move_annotations_task(self, patch):
        return patch("h.services.url_migration.move_annotations")
```